mailpoet/lib/Config/Initializer.php

```diff
@@ -253,11 +255,17 @@ public function init() {
       'register',
     ]);
 
+    WPFunctions::get()->addAction('admin_init', [
+      $this,
+      'afterPluginActivation',
+    ]);
+
     $this->hooks->initEarlyHooks();
   }
 
   public function runActivator() {
     try {
+      $this->wpFunctions->addOption(self::PLUGIN_ACTIVATED, true);
```

Comment:
Feels like a bit too soon to set plugin activated, maybe `Activator::processActivate` is a better place? and since we are deleting this option in `afterPluginActivation`, wouldn't `WPFunctions::doAction` be a better replacement, as in `afterPluginActivation` the `mailpoet_plugin_activated` option finally is deleted while the plugin is still active, I'm afraid it may get confusing in future, what do you think?

Reply:
> Feels like a bit too soon to set plugin activated, maybe Activator::processActivate is a better place?

No, I believe this is okay.
The action is not executed here (on registerActivationHook); it is done in [the admin_init](https://github.com/mailpoet/mailpoet/blob/47ea982eb44984dddacfe7398a938cf7561b697f/mailpoet/lib/Config/Initializer.php#L258-L261) hook.


> since we are deleting this option in afterPluginActivation, wouldn't WPFunctions::doAction be a better replacement, as in afterPluginActivation the mailpoet_plugin_activated option finally is deleted while the plugin is still active, I'm afraid it may get confusing in future, what do you think?

The `mailpoet_plugin_activated` option acts as a flag here to indicate a plugin activation action.

From [WP Docs](https://developer.wordpress.org/reference/functions/register_activation_hook/#process-flow), it is recommended not to perform the type of action we are trying to do here (redirect to another page) on `registerActivationHook` (which includes `Activator::processActivate` as well).

> If you are interested in doing something just after a plugin has been activated it is important to note that the hook process performs an instant redirect after it fires. So it is impossible to use [add_action()](https://developer.wordpress.org/reference/functions/add_action/) or [add_filter()](https://developer.wordpress.org/reference/functions/add_filter/) type calls until the redirect has occurred

While it's possible to use `WPFunctions::doAction`, It would not work for our case as the `registerActivationHook` requires the  instant redirect to be successful. That would not happen if we hook to `WPFunctions::doAction` on `registerActivationHook` and redirect the user to the landing page.

This process is more of a walkaround for getting it to work correctly.

> I'm afraid it may get confusing in future

Would you be okay with linking to this conversation and the [WP Docs](https://developer.wordpress.org/reference/functions/register_activation_hook/#process-flow) in a comment on the code?

Reply:
Thanks for the elaborate explanation @triple0t, I think adding a link to the conversion on `WP Docs` sounds like a good idea to reduce future confusion.